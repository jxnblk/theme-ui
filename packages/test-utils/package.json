{
  "name": "@theme-ui/test-utils",
  "version": "0.11.1",
  "private": true,
  "license": "MIT",
  "repository": "system-ui/theme-ui",
  "source": "src/index.ts",
  "main": "dist/theme-ui-test-utils.cjs.js",
  "module": "dist/theme-ui-test-utils.esm.js",
  "types": "dist/theme-ui-test-utils.cjs.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "@testing-library/react": "^12.1.1",
    "conditional-type-checks": "^1.0.5",
    "react-test-renderer": "17.0.2",
    "ts-snippet": "^5.0.0"
  }
}
