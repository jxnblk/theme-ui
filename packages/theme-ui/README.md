<img src='https://raw.githubusercontent.com/system-ui/theme-ui/master/packages/docs/static/icon.png' width='64' heigh='64' />

# Theme UI

Build consistent, themeable React UIs based on design system constraints and design tokens

[![GitHub][github-badge]][github]
[![Build Status][circleci-badge]][circleci]
[![Version][version]][npm]
![MIT License][license]
[![system-ui/theme][system-ui-badge]](https://system-ui.com/theme)
![][size]

https://theme-ui.com

[github]: https://github.com/system-ui/theme-ui
[github-badge]: https://flat.badgen.net/badge/-/github?icon=github&label
[circleci]: https://circleci.com/gh/system-ui/theme-ui
[circleci-badge]: https://flat.badgen.net/circleci/github/system-ui/theme-ui/master
[version]: https://flat.badgen.net/npm/v/theme-ui
[npm]: https://npmjs.com/package/theme-ui
[license]: https://flat.badgen.net/badge/license/MIT/blue
[system-ui-badge]: https://flat.badgen.net/badge/system-ui/theme/black
[size]: https://flat.badgen.net/bundlephobia/minzip/theme-ui

- Style your application consistently with a global theme object and customizable design tokens
- Style [MDX][] content with themes
- Follows the [Theme Specification][] for interoperability with other libraries
- Style elements directly with the `css` prop and [@styled-system/css][]
- Works with or without custom styled components and [Styled System][]
- First-class support for [Typography.js][] themes
- First-class support for dark mode and other color modes
- Use contextual theming for nested stylistic changes
- Primitive page layout components
- Keep styles isolated with [Emotion][]

[emotion]: https://emotion.sh
[mdx]: https://mdxjs.com
[styled system]: https://styled-system.com
[@styled-system/css]: https://styled-system.com/css
[theme specification]: https://system-ui.com/theme
[typography.js]: https://github.com/KyleAMathews/typography.js

## Getting Started

```sh
npm i theme-ui
```

Theme UI also requires the following dependencies

```sh
npm i @emotion/core @emotion/styled @mdx-js/react
```

Wrap your application with the `ThemeProvider` component and pass in a custom `theme` object.

```jsx
// basic usage
import React from 'react'
import { ThemeProvider } from 'theme-ui'
import theme from './theme'

export default props => (
  <ThemeProvider theme={theme}>{props.children}</ThemeProvider>
)
```

The `theme` object should follow the System UI [Theme Specification][],
which lets you define custom color palettes, typographic scales, fonts, and more.

```js
// example theme.js
export default {
  fonts: {
    body: 'system-ui, sans-serif',
    heading: '"Avenir Next", sans-serif',
    monospace: 'Menlo, monospace',
  },
  colors: {
    text: '#000',
    background: '#fff',
    primary: '#33e',
  },
}
```

## `css` prop

Use the `css` prop throughout your application, along with the `css` utility to pick up values from the theme.
Using the `css` utility means that
color and other values can reference values defined in `theme` object.

```jsx
import React from 'react'
import { css } from 'theme-ui'

export default () => (
  <div
    css={css({
      fontWeight: 'bold',
      fontSize: 4, // picks up value from `theme.fontSizes[4]`
      color: 'primary', // picks up value from `theme.colors.primary`
    })}
  >
    Hello
  </div>
)
```

Read more about the `css` utility in the [@styled-system/css](https://styled-system.com/css/) docs.

## Responsive styles

The `css` utility also supports using arrays as values to change properties responsively with a mobile-first approach.

```jsx
import React from 'react'
import { css } from 'theme-ui'

export default props => (
  <div
    css={css({
      // applies width 100% to all viewport widths,
      // width 50% above the first breakpoint,
      // and 25% above the next breakpoint
      width: ['100%', '50%', '25%'],
    })}
  />
)
```

## `jsx` pragma

To use values from the `theme` object without importing the `css` utility, use the custom `jsx` pragma, which will automatically convert Theme UI `css` prop values.

```jsx
/** @jsx jsx */
import { jsx } from 'theme-ui'

export default props => (
  <div
    css={{
      fontSize: 4,
      color: 'primary',
      bg: 'lightgray',
    }}
  />
)
```

Read more in the [custom pragma docs](https://theme-ui.com/custom-pragma).

## MDX Components

Use the `components` prop to add components to MDX scope.
This can be used to introduce new behavior to markdown elements in MDX, such as adding linked headings or live-editable code examples.
Internally, the `ThemeProvider` is a combination of `MDXProvider` and Emotion's `ThemeProvider`.

```jsx
// with mdx components
import React from 'react'
import { ThemeProvider } from 'theme-ui'
import mdxComponents from './mdx-components'
import theme from './theme'

export default props => (
  <ThemeProvider components={mdxComponents} theme={theme}>
    {props.children}
  </ThemeProvider>
)
```

## `theme.styles`

The MDX components can also be styled via the `theme.styles` object.
This can be used as a mechanism to pass in fully-baked themes and typographic styles to MDX content.

```js
// example theme
export default {
  colors: {
    primary: '#33e',
  },
  styles: {
    // this styles child MDX `<h1>` components
    h1: {
      fontSize: 32,
      // this value comes from the `color` object
      color: 'primary',
    },
  },
}
```

## Styled components

These components' styles can also be consumed _outside_ of an MDX doc with the `Styled` component.
Note that these are only _styled_ using the same `theme.styles` object and _not_ the same components passed to the `ThemeProvider` context.

```jsx
import React from 'react'
import { Styled } from 'theme-ui'

export default props => (
  <Styled.div>
    <Styled.h1>Hello</Styled.h1>
  </Styled.div>
)
```

To change the underlying component in `Styled`, use the `as` prop.

```jsx
<Styled.a as={Link} to="/">
  Home
</Styled.a>
```

## Layout Components

Theme UI includes several components for creating page layouts.

- `Layout`: sets a flex column with 100vh min-height
- `Header`: flexbox row
- `Footer`: flexbox row
- `Main`: flex auto container for pushing the Footer to the bottom of the viewport
- `Container`: max-width, centered container

```jsx
import React from 'react'
import { Layout, Header, Main, Container, Footer } from 'theme-ui'

export default props => (
  <Layout>
    <Header>Hello</Header>
    <Main>
      <Container>{props.children}</Container>
    </Main>
    <Footer>Â© 2019</Footer>
  </Layout>
)
```

## Box & Flex

The `Box` & `Flex` layout components are similar to the ones found in [Rebass](https://rebassjs.org), but are built with Emotion and `@styled-system/css`.

```jsx
import React from 'react'
import { Flex, Box } from 'theme-ui'

export default props => (
  <Flex flexWrap="wrap">
    <Box width={[1, 1 / 2]} />
    <Box width={[1, 1 / 2]} />
  </Flex>
)
```

## More Documentation

- [Theming](https://theme-ui.com/theming)
- [Layout](https://theme-ui.com/layout)
- [Color Modes](https://theme-ui.com/color-modes)
- [Gatsby Plugin](https://theme-ui.com/gatsby-plugin)
- [Custom Pragma](https://theme-ui.com/custom-pragma)
- [Typography.js](https://theme-ui.com/typography)

[typography demo]: https://theme-ui.com/typography
[demo]: https://theme-ui.com/demo
[emotion plugins]: https://github.com/emotion-js/emotion/pull/1299

MIT License
