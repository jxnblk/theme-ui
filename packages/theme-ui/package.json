{
  "name": "theme-ui",
  "version": "0.2.52",
  "description": "Build consistent, themeable React UIs based on design system constraints",
  "main": "dist/index.js",
  "module": "dist/index.esm.js",
  "author": "Brent Jackson <jxnblk@gmail.com>",
  "license": "MIT",
  "repository": "system-ui/theme-ui",
  "scripts": {
    "prepare": "microbundle --no-compress",
    "watch": "microbundle watch --no-compress"
  },
  "peerDependencies": {
    "@emotion/core": "^10.0.0",
    "@mdx-js/react": "^1.0.0",
    "react": "^16.8.0"
  },
  "dependencies": {
    "@emotion/is-prop-valid": "^0.8.1",
    "@theme-ui/css": "^0.2.48",
    "deepmerge": "^4.0.0"
  },
  "devDependencies": {
    "@emotion/core": "^10.0.0",
    "@mdx-js/react": "^1.0.0"
  },
  "keywords": [
    "theme-ui",
    "emotion",
    "mdx",
    "css",
    "styles",
    "css-in-js"
  ],
  "gitHead": "bfd026cae085f377ca537de897dc43233d50f5d5"
}
