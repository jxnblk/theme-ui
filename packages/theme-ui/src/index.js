export { css } from '@styled-system/css'
export { jsx } from './jsx'
export { ThemeProvider } from './core'
export { Context, useThemeUI } from './context'
export { ColorMode, useColorMode } from './color-modes'
export { Styled } from './components'
export { Box, Flex, Layout, Header, Main, Container, Footer } from './layout'

// DEPRECATED
export {
  ThemeProvider as ColorModeProvider,
  ThemeProvider as ComponentProvider,
} from './core'
