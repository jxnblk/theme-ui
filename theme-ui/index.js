import React, { useContext } from 'react'
import { jsx, Global } from '@emotion/core'
import styled from '@emotion/styled'
import { ThemeProvider as EmotionProvider } from 'emotion-theming'
import { MDXProvider } from '@mdx-js/react'
import css from '@styled-system/css'
import merge from 'lodash.merge'
import get from 'lodash.get'

export { css } from '@styled-system/css'

const tags = [
  'p',
  'a',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'img',
  'pre',
  'code',
  'ol',
  'ul',
  'li',
  'blockquote',
  'hr',
  'em',
  'table',
  'tr',
  'th',
  'td',
  'em',
  'strong',
  'delete',
  // mdx
  'inlineCode',
  'thematicBreak',
  // other
  'div',
  // theme-ui
  'root',
]

const aliases = {
  inlineCode: 'code',
  thematicBreak: 'hr',
  root: 'div',
}

const alias = n => aliases[n] || n

const themed = key => theme => css(get(theme, `styles.${key}`))(theme)

export const Styled = React.forwardRef(({
  tag = 'div',  // tag is used as a key in theme.styles
  as,           // as replaces the rendered element type
  ...props
}, ref) =>
  jsx(as || alias(tag), {
    ...props,
    ref,
    css: themed(tag)
  })
)

const components = {}
tags.forEach(tag => {
  components[tag] = styled(alias(tag))(props => themed(tag)(props.theme))
  Styled[tag] = React.forwardRef((props, ref) =>
    jsx(Styled, { ref, tag, ...props })
  )
})

export const Context = React.createContext({
  theme: {},
  components,
})

const createComponents = (components = {}) => {
  const next = {}
  Object.keys(components).forEach(key => {
    next[key] = styled(components[key])(props => themed(key)(props.theme))
  })
  return next
}

export const ComponentProvider = ({
  theme,
  components,
  ...props
}) => {
  const outer = useContext(Context)
  const context = merge({}, outer, {
    theme,
    components: createComponents(components),
  })

  return (
    jsx(EmotionProvider, { theme: context.theme },
      jsx(MDXProvider, { components: context.components },
        jsx(Context.Provider, {
          value: context,
          children: props.children
        })
      )
    )
  )
}

export const Reset = props => React.createElement(Global, {
  styles: {
    '*': {
      boxSizing: 'border-box',
    },
    body: {
      margin: 0,
    },
  }
})

////////
// Potentially remove this
export const useComponents = () => {
  const context = useContext(Context)
  return context.components
}

// This could be removed, but keeping the theme-ui API intact
export const ThemeProvider = props =>
  jsx(ComponentProvider, props)
