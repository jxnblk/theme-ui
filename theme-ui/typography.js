// based on https://github.com/jxnblk/typography-mdx
import Typography from '@jxnblk/typography'
import merge from 'lodash.merge'
import cssWhat from 'css-what'

const isMedia = str => /^@/.test(str)

const parseSelectors = styles => {
  const parsed = {}
  for (const key in styles) {
    // todo: handle media queries
    if (isMedia(key)) continue
    const selectors = key.split(',') //cssWhat(key)
    selectors.forEach(selector => {
      const [ parent, ...children ] = cssWhat(selector)[0]
      if (parent.type === 'universal') return
      if (parent.type !== 'tag') {
        // todo...
        // console.log('!TAG', selector, parent.type)
        return
      }

      if (children.length) {
        const sub = selector.replace(parent.name, '&')
        parsed[parent.name] = merge({}, parsed[parent.name], {
          [sub]: styles[key]
        })
      } else {
        parsed[parent.name] = merge({}, parsed[parent.name], styles[key])
      }
    })
  }
  for (const key in parsed) {
    if (key === 'html' || key === 'body') {
      // todo: what should this be named?
      parsed.root = merge({}, parsed.root, parsed[key])
    }
  }
  return parsed
}

export const toStyles = (theme) => {
  const typography = new Typography({
    ...theme,
    output: 'components',
    rhythmUnit: 'px',
    includeNormalize: false,
  })
  const json = typography.toJSON()

  // still need this for themes that add custom styles
  const styles = parseSelectors(json)

  typography.styles = styles
  return typography
}

export default toStyles
