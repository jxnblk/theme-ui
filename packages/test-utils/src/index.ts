import renderer from 'react-test-renderer'
import * as tsSnippet from 'ts-snippet'

export * from '@testing-library/react'

export type {
  Assert,
  AssertFalse,
  AssertTrue,
  Has,
  IsAny,
  IsExact,
  IsNever,
  IsNullable,
  IsUnknown,
  NotHas,
} from 'conditional-type-checks'

export const renderJSON = (
  el: Parameters<typeof renderer.create>[0]
): renderer.ReactTestRendererJSON | null => {
  const json = renderer.create(el).toJSON()
  if (Array.isArray(json)) {
    return json[0]
  }

  return json
}

/**
 * @param setupCode imports, common code
 * @example
 * const expectSnippet = expecter(`
 *  import { add } from './made-up/add'
 * `);
 *
 * expectSnippet(`
 *   const x = add(10, 20)
 * `).toInfer('x', 'number')
 */
export const expecter = (
  setupCode: string,
  options: { jsx?: boolean } = {}
) => {
  options = { jsx: true, ...options }

  const fileName = 'snippet.tsx'

  return (code: string) => {
    const compiler = new tsSnippet.Compiler({
      // same settings as our root tsconfig
      resolveJsonModule: true,
      esModuleInterop: true,
      moduleResolution: 'node',
      strict: true,
      jsx: 'react',
      isolatedModules: true,

      // for better error message snapshots
      noErrorTruncation: true,

      types: ['node'],
    })

    const snippets = tsSnippet.snippet(
      {
        [fileName]: `
        ${
          options.jsx
            ? `
            /** @jsx jsx **/
            import { jsx } from './packages/theme-ui'
          `
            : ''
        }
        ${setupCode}
        ${code}
      `,
      },
      compiler
    )

    return snippets.expect(fileName)
  }
}
