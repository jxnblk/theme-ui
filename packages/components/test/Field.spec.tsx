/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderJSON } from '@theme-ui/test-utils'
import { ThemeUIProvider } from 'theme-ui'

import { Field } from '..'

import { theme } from './__test-utils__'

describe('Field', () => {
  test('renders', () => {
    const json = renderJSON(
      <ThemeUIProvider theme={theme}>
        <Field />
      </ThemeUIProvider>
    )
    expect(json).toMatchSnapshot()
  })
  test('renders with id prop', () => {
    const json = renderJSON(
      <ThemeUIProvider theme={theme}>
        <Field id="test-field" />
      </ThemeUIProvider>
    )
    expect(json).toMatchSnapshot()
  })
  test('containerSx and sx', () => {
    const json = renderJSON(
      <ThemeUIProvider theme={theme}>
        <Field label="Name" sx={{ color: 'primary' }} containerSx={{ my: 4 }} />
      </ThemeUIProvider>
    )
    expect(json).toMatchSnapshot()
  })
})
