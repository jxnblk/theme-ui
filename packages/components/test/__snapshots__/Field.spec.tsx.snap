// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field containerSx and sx 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 32px;
  margin-bottom: 32px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  background-color: transparent;
  color: primary;
}

.emotion-2:autofill,
.emotion-2:autofill:hover,
.emotion-2:autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-2:autofill:first-line,
.emotion-2:autofill:hover:first-line,
.emotion-2:autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-2:-webkit-autofill,
.emotion-2:-webkit-autofill:hover,
.emotion-2:-webkit-autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-2:-webkit-autofill:first-line,
.emotion-2:-webkit-autofill:hover:first-line,
.emotion-2:-webkit-autofill:focus:first-line {
  font-size: 1rem;
}

<div
  className="emotion-0"
>
  <label
    className="emotion-1"
  >
    Name
  </label>
  <input
    className="emotion-2"
  />
</div>
`;

exports[`Field renders 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  background-color: transparent;
}

.emotion-2:autofill,
.emotion-2:autofill:hover,
.emotion-2:autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-2:autofill:first-line,
.emotion-2:autofill:hover:first-line,
.emotion-2:autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-2:-webkit-autofill,
.emotion-2:-webkit-autofill:hover,
.emotion-2:-webkit-autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-2:-webkit-autofill:first-line,
.emotion-2:-webkit-autofill:hover:first-line,
.emotion-2:-webkit-autofill:focus:first-line {
  font-size: 1rem;
}

<div
  className="emotion-0"
>
  <label
    className="emotion-1"
  />
  <input
    className="emotion-2"
  />
</div>
`;

exports[`Field renders with id prop 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  background-color: transparent;
}

.emotion-2:autofill,
.emotion-2:autofill:hover,
.emotion-2:autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-2:autofill:first-line,
.emotion-2:autofill:hover:first-line,
.emotion-2:autofill:focus:first-line {
  font-size: 1rem;
}

.emotion-2:-webkit-autofill,
.emotion-2:-webkit-autofill:hover,
.emotion-2:-webkit-autofill:focus {
  box-shadow: inset 0 0 0 1000px var(--theme-ui-input-autofill-bg);
  font-size: inherit;
}

.emotion-2:-webkit-autofill:first-line,
.emotion-2:-webkit-autofill:hover:first-line,
.emotion-2:-webkit-autofill:focus:first-line {
  font-size: 1rem;
}

<div
  className="emotion-0"
>
  <label
    className="emotion-1"
    htmlFor="test-field"
  />
  <input
    className="emotion-2"
    id="test-field"
  />
</div>
`;
